File: app/src/main/java/com/tacocat/catwithumbrella/WellcomeActivity.java
package com.tacocat.catwithumbrella;

import android.Manifest;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.widget.TextView;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.tasks.OnSuccessListener;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class WellcomeActivity extends AppCompatActivity {

    private FusedLocationProviderClient mFusedLocationClient;
    private TextView addressTxtV;
    private TextView weatherTxtV;
    private ApiInterface apiInterface;
    private LocationCallback mLocationCallback;
    private static LocationRequest locationRequest = createLocationRequest();


    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_wellcome);
        addressTxtV = findViewById(R.id.current_location);
        weatherTxtV = findViewById(R.id.weatherTxtV);
        createLocationCallback();
        apiInterface = ApiClient.Companion.getClient().create(ApiInterface.class);
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this);

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION}, 3);
            return;
        }

        setLocation();
    }

    private void createLocationCallback() {
        mLocationCallback = new LocationCallback() {
            @Override
            public void onLocationResult(LocationResult locationResult) {
                Log.d("Tag", "onLocationResult");
                Location location = locationResult.getLastLocation();
                if (location != null) {
                    mFusedLocationClient.removeLocationUpdates(mLocationCallback);

                    Address address = getLocationName(location);
                    if (address != null) {
                        addressTxtV.setText(address.getAddressLine(0));
                        getWeather(address.getLatitude(), address.getLongitude());
                    }
                }
            }
        };
    }

    private void setLocation() {
        Log.d("Tag", "setLocation called");
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        mFusedLocationClient.requestLocationUpdates(locationRequest, mLocationCallback, null );
    }

    private Address getLocationName(Location location) {
        Geocoder geocoder = new Geocoder(this, Locale.getDefault());
        try {
            List<Address> addresses = geocoder.getFromLocation(
                    location.getLatitude(),
                    location.getLongitude(),
                    1);
            return addresses.get(0);
        } catch (IOException e) {
            e.printStackTrace();
        }
          return null;
    }

    public void onRequestPermissionsResult(int requestCode, String[] permissions,
                                            int[] grantResults){
        setLocation();
    }

    public void getWeather(double lat, double lng) {
        Log.d("Tag", "getWeather called");
        Call<WeatherResponse> call = apiInterface.getWeather(lat, lng, "a9f07b3e58aa83b04d6299075004c1ce");
        call.enqueue(new Callback<WeatherResponse>() {
            @Override
            public void onResponse(Call<WeatherResponse> call, Response<WeatherResponse> response) {
                Log.d("Tag", "getWeather response: " + response.body());
                WeatherResponse weatherResponse = response.body();
                WeatherResponse.WeatherInfo weatherInfo = weatherResponse.getWeatherInfoList().get(0);
                WeatherResponse.Weather weather1 = weatherInfo.getWeatherList().get(0);
                weatherTxtV.setText(weather1.getMain());
            }

            @Override
            public void onFailure(Call<WeatherResponse> call, Throwable t) {
               call.cancel();
            }
        });
    }

    protected static LocationRequest createLocationRequest() {
        LocationRequest mLocationRequest = new LocationRequest();
        mLocationRequest.setInterval(10000);
        mLocationRequest.setFastestInterval(5000);
        mLocationRequest.setPriority(LocationRequest.PRIORITY_LOW_POWER);
        return mLocationRequest;
    }
}

File: app/src/main/java/com/tacocat/catwithumbrella/WellcomeActivity.kt
package com.tacocat.catwithumbrella

import android.Manifest
import android.content.pm.PackageManager
import android.location.Address
import android.location.Geocoder
import android.location.Location
import android.os.Build
import android.os.Bundle
import android.support.annotation.RequiresApi
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.transition.TransitionManager
import android.util.Log
import android.view.View
import com.google.android.gms.location.*
import com.tacocat.catwithumbrella.R.id.*
import com.tacocat.catwithumbrella.networking.ApiClient
import com.tacocat.catwithumbrella.networking.ApiInterface
import com.tacocat.catwithumbrella.networking.WeatherResponse
import kotlinx.android.synthetic.main.activity_wellcome.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException
import java.util.*

class WellcomeActivity : AppCompatActivity() {

    private var mFusedLocationClient: FusedLocationProviderClient? = null
    private var apiInterface: ApiInterface? = null
    private var mLocationCallback: LocationCallback? = null


    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_wellcome)
        Util.sheduleJob(applicationContext)

        locationTxtV.visibility = View.GONE
        weatherTxtV.visibility = View.GONE
        dayTxtV.visibility = View.GONE

        createLocationCallback()

        apiInterface = ApiClient.getClient().create(ApiInterface::class.java)
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this)

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
                != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION)
                != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.ACCESS_FINE_LOCATION,
                    Manifest.permission.ACCESS_COARSE_LOCATION), 3)
            return
        }

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
            // Permission is not granted
            // Request for permission
            ActivityCompat.requestPermissions(this,
                    arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                   1);
        }

            setLocation()
    }

    private fun createLocationCallback() {
        mLocationCallback = object : LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult?) {
                Log.d("Tag", "onLocationResult")
                val location = locationResult!!.lastLocation
                if (location != null) {
                    mFusedLocationClient!!.removeLocationUpdates(mLocationCallback!!)

                    setLocationText(location)
                }
            }
        }
    }

    private fun setLocationText(location: Location?) {
        val address = getLocationName(location)
        if (address != null) {

            var arr = address.getAddressLine(0).split(",")
            var textArr = arr.takeLast(2)
            locationTxtV.text = "${textArr[0]}, ${textArr[1]}"
            TransitionManager.beginDelayedTransition(weatherContainer)
            locationTxtV.visibility = View.VISIBLE

            getWeather(address.latitude, address.longitude)
        }
    }

    private fun setLocation() {
        Log.d("Tag", "setLocation called")
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return
        }
        mFusedLocationClient!!.requestLocationUpdates(locationRequest, mLocationCallback!!, null)
    }

    private fun getLocationName(location: Location?): Address? {
        val geocoder = Geocoder(this, Locale.getDefault())
        try {
            val addresses = geocoder.getFromLocation(
                    location!!.latitude,
                    location.longitude,
                    1)
            Log.d("Address", addresses.toString() + "")
            return addresses[0]
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        setLocation()
    }

    fun getWeather(lat: Double, lng: Double) {
        Log.d("Tag", "getWeather called")
        val call = apiInterface!!.getWeather("metric", lat, lng, "a9f07b3e58aa83b04d6299075004c1ce")
        call.enqueue(object : Callback<WeatherResponse> {
            override fun onResponse(call: Call<WeatherResponse>, response: Response<WeatherResponse>) {
                onWeatherRetrieved(response)
            }

            override fun onFailure(call: Call<WeatherResponse>, t: Throwable) {
                call.cancel()
            }
        })
    }

    private fun onWeatherRetrieved(response: Response<WeatherResponse>) {
        Log.d("Tag", "getWeather response: " + response.body())

        weatherContainer.setOnClickListener{
            onWeatherRetrieved(response)
        }

        weatherTxtV.visibility = View.GONE
        dayTxtV.visibility = View.GONE

        val weatherResponse = response.body()
        showToday(weatherResponse)


        TransitionManager.beginDelayedTransition(weatherContainer)
        weatherTxtV.visibility = View.VISIBLE
        dayTxtV.visibility = View.VISIBLE

        dayTxtV.postDelayed({
            showTomorrow(weatherResponse)
        }, 3000)

        dayTxtV.postDelayed({
            showWeekend(weatherResponse)
        }, 7000)
    }

    private fun showToday(weatherResponse: WeatherResponse?) {
        val now = Calendar.getInstance()
        dayTxtV.text = "Today"
        setWeatherText(weatherResponse, now)
    }

    private fun showWeekend(weatherResponse: WeatherResponse?) {
        dayTxtV.visibility = View.GONE
        weatherTxtV.visibility = View.GONE


        dayTxtV.text = "Weekend"

        val calWeather = Calendar.getInstance()
        weatherTxtV.text = "NOT SURE"
        var willRain = false
        var found = false
        weatherResponse!!.list!!.forEach {
            calWeather.timeInMillis = it.dt * 1000
            Log.d("t", calWeather.get(Calendar.DAY_OF_WEEK).toString() + " weather: " + it.weather!![0].main)
            if (calWeather.get(Calendar.DAY_OF_WEEK) == Calendar.SUNDAY || calWeather.get(Calendar.DAY_OF_WEEK) == Calendar.SATURDAY) {
                if (it.weather!![0].main.equals("Rain", true)) {
                    willRain = true
                }
                found = true
            }
        }

        if(willRain){
            weatherTxtV.text = "YES"
        }else if(found){
            weatherTxtV.text = "NO"
        }

        TransitionManager.beginDelayedTransition(weatherContainer)
        dayTxtV.visibility = View.VISIBLE
        weatherTxtV.visibility = View.VISIBLE
    }

    private fun setWeatherText(weatherResponse: WeatherResponse?, dateToCompare: Calendar) {
        val calWeather = Calendar.getInstance()
        weatherTxtV.text = "NO"
        weatherResponse!!.list!!.forEach {
            calWeather.timeInMillis = it.dt * 1000
            if (dateToCompare.get(Calendar.DATE) == calWeather.get(Calendar.DATE)) {
                if (it.weather!![0].main.equals("Rain", true)) {
                    weatherTxtV.text = "YES"
                    return
                }
            }
        }
    }



    private fun showTomorrow(weatherResponse: WeatherResponse) {
        dayTxtV.visibility = View.GONE
        weatherTxtV.visibility = View.GONE


        dayTxtV.text = "Tomorrow"
        var tomorrow= Calendar.getInstance()
        tomorrow.add(Calendar.DATE, 1)
        setWeatherText(weatherResponse, tomorrow)

        TransitionManager.beginDelayedTransition(weatherContainer)
        dayTxtV.visibility = View.VISIBLE
        weatherTxtV.visibility = View.VISIBLE
    }


    companion object {
        private val locationRequest = createLocationRequest()

        protected fun createLocationRequest(): LocationRequest {
            val mLocationRequest = LocationRequest()
            mLocationRequest.interval = 10000
            mLocationRequest.fastestInterval = 5000
            mLocationRequest.priority = LocationRequest.PRIORITY_LOW_POWER
            return mLocationRequest
        }
    }
}
